Godot 4: Switching to GLES3 - CanvasShaderGLES3 Fragment Shader compilation failed.

### Godot version

4.0 alpha 16

### System information

Windows 11 - GLES3

### Issue description

When switching from Vulkan to GLES3, any shader that has been modified to use Vulkan, will no longer compile under GLES3, failing with the error:

```
E 0:00:03:0897   _display_error_with_code: CanvasShaderGLES3: Fragment shader compilation failed:
0(390) : error C1503: undefined variable "canvas_data"

  <C++ Source>   drivers\gles3\shader_gles3.cpp:216 @ _display_error_with_code()
```

Code from First Shader (Have Multiple):
```
1: #version 330
2: #define USE_GLES_OVER_GL
3: #define MATERIAL_UNIFORMS_USED
4: #define FRAGMENT_CODE_USED
5: 
6: #define MAX_GLOBAL_SHADER_UNIFORMS 256
7: #define MAX_LIGHTS 512
8: #define MAX_DRAW_DATA_INSTANCES 512
9: 
10: #define SCREEN_UV_USED
11: #define COLOR_USED
12: 
13: precision highp float;
14: precision highp int;
15: 
16: 
17: #define MAX_LIGHTS_PER_ITEM uint(16)
18: 
19: #define M_PI 3.14159265359
20: 
21: #define SDF_MAX_LENGTH 16384.0
22: 
23: //1 means enabled, 2+ means trails in use
24: #define FLAGS_INSTANCING_MASK uint(0x7F)
25: #define FLAGS_INSTANCING_HAS_COLORS uint(1 << 7)
26: #define FLAGS_INSTANCING_HAS_CUSTOM_DATA uint(1 << 8)
27: 
28: #define FLAGS_CLIP_RECT_UV uint(1 << 9)
29: #define FLAGS_TRANSPOSE_RECT uint(1 << 10)
30: #define FLAGS_USING_LIGHT_MASK uint(1 << 11)
31: #define FLAGS_NINEPACH_DRAW_CENTER uint(1 << 12)
32: #define FLAGS_USING_PARTICLES uint(1 << 13)
33: 
34: #define FLAGS_NINEPATCH_H_MODE_SHIFT 16
35: #define FLAGS_NINEPATCH_V_MODE_SHIFT 18
36: 
37: #define FLAGS_LIGHT_COUNT_SHIFT 20
38: 
39: #define FLAGS_DEFAULT_NORMAL_MAP_USED uint(1 << 26)
40: #define FLAGS_DEFAULT_SPECULAR_MAP_USED uint(1 << 27)
41: 
42: #define FLAGS_USE_MSDF uint(1 << 28)
43: #define FLAGS_USE_LCD uint(1 << 29)
44: 
45: // must be always 128 bytes long
46: struct DrawData {
47: 	vec2 world_x;
48: 	vec2 world_y;
49: 	vec2 world_ofs;
50: 	vec2 color_texture_pixel_size;
51: #ifdef USE_PRIMITIVE
52: 	vec2 point_a;
53: 	vec2 point_b;
54: 	vec2 point_c;
55: 	vec2 uv_a;
56: 	vec2 uv_b;
57: 	vec2 uv_c;
58: 	uint color_a_rg;
59: 	uint color_a_ba;
60: 	uint color_b_rg;
61: 	uint color_b_ba;
62: 	uint color_c_rg;
63: 	uint color_c_ba;
64: #else
65: 	vec4 modulation;
66: 	vec4 ninepatch_margins;
67: 	vec4 dst_rect; //for built-in rect and UV
68: 	vec4 src_rect;
69: 	uint pad;
70: 	uint pad2;
71: #endif
72: 	uint flags;
73: 	uint specular_shininess;
74: 	uvec4 lights;
75: };
76: 
77: layout(std140) uniform GlobalShaderUniformData { //ubo:1
78: 	vec4 global_shader_uniforms[MAX_GLOBAL_SHADER_UNIFORMS];
79: };
80: 
81: layout(std140) uniform CanvasData { //ubo:0
82: 	mat4 canvas_transform;
83: 	mat4 screen_transform;
84: 	mat4 canvas_normal_transform;
85: 	vec4 canvas_modulation;
86: 	vec2 screen_pixel_size;
87: 	float time;
88: 	bool use_pixel_snap;
89: 
90: 	vec4 sdf_to_tex;
91: 	vec2 screen_to_sdf;
92: 	vec2 sdf_to_screen;
93: 
94: 	uint directional_light_count;
95: 	float tex_to_sdf;
96: 	uint pad1;
97: 	uint pad2;
98: };
99: 
100: #define LIGHT_FLAGS_BLEND_MASK uint(3 << 16)
101: #define LIGHT_FLAGS_BLEND_MODE_ADD uint(0 << 16)
102: #define LIGHT_FLAGS_BLEND_MODE_SUB uint(1 << 16)
103: #define LIGHT_FLAGS_BLEND_MODE_MIX uint(2 << 16)
104: #define LIGHT_FLAGS_BLEND_MODE_MASK uint(3 << 16)
105: #define LIGHT_FLAGS_HAS_SHADOW uint(1 << 20)
106: #define LIGHT_FLAGS_FILTER_SHIFT 22
107: #define LIGHT_FLAGS_FILTER_MASK uint(3 << 22)
108: #define LIGHT_FLAGS_SHADOW_NEAREST uint(0 << 22)
109: #define LIGHT_FLAGS_SHADOW_PCF5 uint(1 << 22)
110: #define LIGHT_FLAGS_SHADOW_PCF13 uint(2 << 22)
111: 
112: layout(std140) uniform DrawDataInstances { //ubo:3
113: 
114: 	DrawData draw_data[MAX_DRAW_DATA_INSTANCES];
115: };
116: 
117: #ifdef USE_GLES_OVER_GL
118: // Floating point pack/unpack functions are part of the GLSL ES 300 specification used by web and mobile.
119: uint float2half(uint f) {
120: 	return ((f >> uint(16)) & uint(0x8000)) |
121: 			((((f & uint(0x7f800000)) - uint(0x38000000)) >> uint(13)) & uint(0x7c00)) |
122: 			((f >> uint(13)) & uint(0x03ff));
123: }
124: 
125: uint half2float(uint h) {
126: 	return ((h & uint(0x8000)) << uint(16)) | (((h & uint(0x7c00)) + uint(0x1c000)) << uint(13)) | ((h & uint(0x03ff)) << uint(13));
127: }
128: 
129: uint packHalf2x16(vec2 v) {
130: 	return float2half(floatBitsToUint(v.x)) | float2half(floatBitsToUint(v.y)) << uint(16);
131: }
132: 
133: vec2 unpackHalf2x16(uint v) {
134: 	return vec2(uintBitsToFloat(half2float(v & uint(0xffff))),
135: 			uintBitsToFloat(half2float(v >> uint(16))));
136: }
137: 
138: uint packUnorm2x16(vec2 v) {
139: 	uvec2 uv = uvec2(round(clamp(v, vec2(0.0), vec2(1.0)) * 65535.0));
140: 	return uv.x | uv.y << uint(16);
141: }
142: 
143: vec2 unpackUnorm2x16(uint p) {
144: 	return vec2(float(p & uint(0xffff)), float(p >> uint(16))) * 0.000015259021; // 1.0 / 65535.0 optimization
145: }
146: 
147: uint packSnorm2x16(vec2 v) {
148: 	uvec2 uv = uvec2(round(clamp(v, vec2(-1.0), vec2(1.0)) * 32767.0) + 32767.0);
149: 	return uv.x | uv.y << uint(16);
150: }
151: 
152: vec2 unpackSnorm2x16(uint p) {
153: 	vec2 v = vec2(float(p & uint(0xffff)), float(p >> uint(16)));
154: 	return clamp((v - 32767.0) * vec2(0.00003051851), vec2(-1.0), vec2(1.0));
155: }
156: #endif
157: 
158: uint packUnorm4x8(vec4 v) {
159: 	uvec4 uv = uvec4(round(clamp(v, vec4(0.0), vec4(1.0)) * 255.0));
160: 	return uv.x | (uv.y << uint(8)) | (uv.z << uint(16)) | (uv.w << uint(24));
161: }
162: 
163: vec4 unpackUnorm4x8(uint p) {
164: 	return vec4(float(p & uint(0xff)), float((p >> uint(8)) & uint(0xff)), float((p >> uint(16)) & uint(0xff)), float(p >> uint(24))) * 0.00392156862; // 1.0 / 255.0
165: }
166: 
167: uint packSnorm4x8(vec4 v) {
168: 	uvec4 uv = uvec4(round(clamp(v, vec4(-1.0), vec4(1.0)) * 127.0) + 127.0);
169: 	return uv.x | uv.y << uint(8) | uv.z << uint(16) | uv.w << uint(24);
170: }
171: 
172: vec4 unpackSnorm4x8(uint p) {
173: 	vec4 v = vec4(float(p & uint(0xff)), float((p >> uint(8)) & uint(0xff)), float((p >> uint(16)) & uint(0xff)), float(p >> uint(24)));
174: 	return clamp((v - vec4(127.0)) * vec4(0.00787401574), vec4(-1.0), vec4(1.0));
175: }
176: 
177: //uniform sampler2D atlas_texture; //texunit:-2
178: //uniform sampler2D shadow_atlas_texture; //texunit:-3
179: uniform sampler2D screen_texture; //texunit:-4
180: uniform sampler2D sdf_texture; //texunit:-5
181: uniform sampler2D normal_texture; //texunit:-6
182: uniform sampler2D specular_texture; //texunit:-7
183: 
184: uniform sampler2D color_texture; //texunit:0
185: 
186: in vec2 uv_interp;
187: in vec4 color_interp;
188: in vec2 vertex_interp;
189: flat in int draw_data_instance;
190: 
191: #ifdef USE_NINEPATCH
192: 
193: in vec2 pixel_size_interp;
194: 
195: #endif
196: 
197: layout(location = 0) out vec4 frag_color;
198: 
199: #ifdef MATERIAL_UNIFORMS_USED
200: layout(std140) uniform MaterialUniforms{
201: //ubo:4
202: 
203: highp float m_speed;
204: highp vec2 m_scale;
205: highp float m_rotate_speed;
206: highp float m_line_width;
207: highp float m_line_size;
208: highp vec4 m_back_color;
209: highp vec4 m_line_color;
210: 
211: };
212: #endif
213: 
214: 
215: float m_get_ratio_scan_line(float m_p)
216: 	{
217: return (max((-sin(((mod(m_p, m_line_size)/m_line_size)*3.1415926535897932384626433833))+m_line_width), 0.0)/m_line_width);	}
218: 
219: float m_hatch(vec2 m_src_uv, float m_time, float m_dir)
220: 	{
221: 		float m_r=((m_time*m_rotate_speed)*m_dir);
222: 		float m_c=cos(m_r);
223: 		float m_s=sin(m_r);
224: 		mat2 m_matr=mat2(vec2(m_c, m_s), vec2(-m_s, m_c));
225: 		vec2 m_uv=(((((m_src_uv-vec2(0.5,0.5))*m_scale)*m_time)*0.5)*m_matr);
226: return clamp((((m_get_ratio_scan_line(m_uv.x)+m_get_ratio_scan_line(m_uv.y))*sin((m_time*3.1415926535897932384626433833)))*4.0), 0.0, 1.0);	}
227: 
228: #ifdef USE_NINEPATCH
229: 
230: float map_ninepatch_axis(float pixel, float draw_size, float tex_pixel_size, float margin_begin, float margin_end, int np_repeat, inout int draw_center) {
231: 	float tex_size = 1.0 / tex_pixel_size;
232: 
233: 	if (pixel < margin_begin) {
234: 		return pixel * tex_pixel_size;
235: 	} else if (pixel >= draw_size - margin_end) {
236: 		return (tex_size - (draw_size - pixel)) * tex_pixel_size;
237: 	} else {
238: 		if (!bool(draw_data[draw_data_instance].flags & FLAGS_NINEPACH_DRAW_CENTER)) {
239: 			draw_center--;
240: 		}
241: 
242: 		// np_repeat is passed as uniform using NinePatchRect::AxisStretchMode enum.
243: 		if (np_repeat == 0) { // Stretch.
244: 			// Convert to ratio.
245: 			float ratio = (pixel - margin_begin) / (draw_size - margin_begin - margin_end);
246: 			// Scale to source texture.
247: 			return (margin_begin + ratio * (tex_size - margin_begin - margin_end)) * tex_pixel_size;
248: 		} else if (np_repeat == 1) { // Tile.
249: 			// Convert to offset.
250: 			float ofs = mod((pixel - margin_begin), tex_size - margin_begin - margin_end);
251: 			// Scale to source texture.
252: 			return (margin_begin + ofs) * tex_pixel_size;
253: 		} else if (np_repeat == 2) { // Tile Fit.
254: 			// Calculate scale.
255: 			float src_area = draw_size - margin_begin - margin_end;
256: 			float dst_area = tex_size - margin_begin - margin_end;
257: 			float scale = max(1.0, floor(src_area / max(dst_area, 0.0000001) + 0.5));
258: 			// Convert to ratio.
259: 			float ratio = (pixel - margin_begin) / src_area;
260: 			ratio = mod(ratio * scale, 1.0);
261: 			// Scale to source texture.
262: 			return (margin_begin + ratio * dst_area) * tex_pixel_size;
263: 		} else { // Shouldn't happen, but silences compiler warning.
264: 			return 0.0;
265: 		}
266: 	}
267: }
268: 
269: #endif
270: 
271: float msdf_median(float r, float g, float b, float a) {
272: 	return min(max(min(r, g), min(max(r, g), b)), a);
273: }
274: 
275: void main() {
276: 	vec4 color = color_interp;
277: 	vec2 uv = uv_interp;
278: 	vec2 vertex = vertex_interp;
279: 
280: #if !defined(USE_ATTRIBUTES) && !defined(USE_PRIMITIVE)
281: 
282: #ifdef USE_NINEPATCH
283: 
284: 	int draw_center = 2;
285: 	uv = vec2(
286: 			map_ninepatch_axis(pixel_size_interp.x, abs(draw_data[draw_data_instance].dst_rect.z), draw_data[draw_data_instance].color_texture_pixel_size.x, draw_data[draw_data_instance].ninepatch_margins.x, draw_data[draw_data_instance].ninepatch_margins.z, int(draw_data[draw_data_instance].flags >> FLAGS_NINEPATCH_H_MODE_SHIFT) & 0x3, draw_center),
287: 			map_ninepatch_axis(pixel_size_interp.y, abs(draw_data[draw_data_instance].dst_rect.w), draw_data[draw_data_instance].color_texture_pixel_size.y, draw_data[draw_data_instance].ninepatch_margins.y, draw_data[draw_data_instance].ninepatch_margins.w, int(draw_data[draw_data_instance].flags >> FLAGS_NINEPATCH_V_MODE_SHIFT) & 0x3, draw_center));
288: 
289: 	if (draw_center == 0) {
290: 		color.a = 0.0;
291: 	}
292: 
293: 	uv = uv * draw_data[draw_data_instance].src_rect.zw + draw_data[draw_data_instance].src_rect.xy; //apply region if needed
294: 
295: #endif
296: 	if (bool(draw_data[draw_data_instance].flags & FLAGS_CLIP_RECT_UV)) {
297: 		uv = clamp(uv, draw_data[draw_data_instance].src_rect.xy, draw_data[draw_data_instance].src_rect.xy + abs(draw_data[draw_data_instance].src_rect.zw));
298: 	}
299: 
300: #endif
301: 
302: #ifndef USE_PRIMITIVE
303: 	if (bool(draw_data[draw_data_instance].flags & FLAGS_USE_MSDF)) {
304: 		float px_range = draw_data[draw_data_instance].ninepatch_margins.x;
305: 		float outline_thickness = draw_data[draw_data_instance].ninepatch_margins.y;
306: 		//float reserved1 = draw_data[draw_data_instance].ninepatch_margins.z;
307: 		//float reserved2 = draw_data[draw_data_instance].ninepatch_margins.w;
308: 
309: 		vec4 msdf_sample = texture(color_texture, uv);
310: 		vec2 msdf_size = vec2(textureSize(color_texture, 0));
311: 		vec2 dest_size = vec2(1.0) / fwidth(uv);
312: 		float px_size = max(0.5 * dot((vec2(px_range) / msdf_size), dest_size), 1.0);
313: 		float d = msdf_median(msdf_sample.r, msdf_sample.g, msdf_sample.b, msdf_sample.a) - 0.5;
314: 
315: 		if (outline_thickness > 0.0) {
316: 			float cr = clamp(outline_thickness, 0.0, px_range / 2.0) / px_range;
317: 			float a = clamp((d + cr) * px_size, 0.0, 1.0);
318: 			color.a = a * color.a;
319: 		} else {
320: 			float a = clamp(d * px_size + 0.5, 0.0, 1.0);
321: 			color.a = a * color.a;
322: 		}
323: 	} else if (bool(draw_data[draw_data_instance].flags & FLAGS_USE_LCD)) {
324: 		vec4 lcd_sample = texture(color_texture, uv);
325: 		if (lcd_sample.a == 1.0) {
326: 			color.rgb = lcd_sample.rgb * color.a;
327: 		} else {
328: 			color = vec4(0.0, 0.0, 0.0, 0.0);
329: 		}
330: 	} else {
331: #else
332: 	{
333: #endif
334: 		color *= texture(color_texture, uv);
335: 	}
336: 
337: 	bool using_light = false;
338: 
339: 	vec3 normal;
340: 
341: #if defined(NORMAL_USED)
342: 	bool normal_used = true;
343: #else
344: 	bool normal_used = false;
345: #endif
346: 
347: 	if (normal_used || (using_light && bool(draw_data[draw_data_instance].flags & FLAGS_DEFAULT_NORMAL_MAP_USED))) {
348: 		normal.xy = texture(normal_texture, uv).xy * vec2(2.0, -2.0) - vec2(1.0, -1.0);
349: 		normal.z = sqrt(1.0 - dot(normal.xy, normal.xy));
350: 		normal_used = true;
351: 	} else {
352: 		normal = vec3(0.0, 0.0, 1.0);
353: 	}
354: 
355: 	vec4 specular_shininess;
356: 
357: #if defined(SPECULAR_SHININESS_USED)
358: 
359: 	bool specular_shininess_used = true;
360: #else
361: 	bool specular_shininess_used = false;
362: #endif
363: 
364: 	if (specular_shininess_used || (using_light && normal_used && bool(draw_data[draw_data_instance].flags & FLAGS_DEFAULT_SPECULAR_MAP_USED))) {
365: 		specular_shininess = texture(specular_texture, uv);
366: 		specular_shininess *= unpackUnorm4x8(draw_data[draw_data_instance].specular_shininess);
367: 		specular_shininess_used = true;
368: 	} else {
369: 		specular_shininess = vec4(1.0);
370: 	}
371: 
372: #if defined(SCREEN_UV_USED)
373: 	vec2 screen_uv = gl_FragCoord.xy * screen_pixel_size;
374: #else
375: 	vec2 screen_uv = vec2(0.0);
376: #endif
377: 
378: 	vec3 light_vertex = vec3(vertex, 0.0);
379: 	vec2 shadow_vertex = vertex;
380: 
381: 	{
382: 		float normal_map_depth = 1.0;
383: 
384: #if defined(NORMAL_MAP_USED)
385: 		vec3 normal_map = vec3(0.0, 0.0, 1.0);
386: 		normal_used = true;
387: #endif
388: 
389: 	{
390: 		float m_time=(mod((canvas_data.time*m_speed), 0.3)/0.3);
391: 		float m_f=clamp((m_hatch(screen_uv, m_time, 1.0)+m_hatch(screen_uv, mod((m_time+0.5), 1.0), -1.0)), 0.0, 1.0);
392: 		color=mix(m_back_color, m_line_color, m_f);
393: 	}
394: 
395: #if defined(NORMAL_MAP_USED)
396: 		normal = mix(vec3(0.0, 0.0, 1.0), normal_map * vec3(2.0, -2.0, 1.0) - vec3(1.0, -1.0, 0.0), normal_map_depth);
397: #endif
398: 	}
399: 
400: #ifdef MODE_LIGHT_ONLY
401: 	color = vec4(0.0);
402: #else
403: 	color *= canvas_modulation;
404: #endif
405: 
406: 	frag_color = color;
407: }
408: 
409:
```

The Shader I believe it is compiling:
```
/*
	電子的十字背景シェーダー by あるる（きのもと 結衣） @arlez80
	Electric Hatch Background Shader by KINOMOTO Yui @arlez80

	MIT License
*/
shader_type canvas_item;

//const float PI = 3.1415926535;

uniform float speed = 0.05;
uniform vec2 scale = vec2( 8.0, 4.5 );
uniform float rotate_speed = 1.1;
uniform float line_width : hint_range( 0.0, 1.0 ) = 0.1;
uniform float line_size = 0.1;

uniform vec4 back_color : source_color = vec4( 0.0, 0.0, 0.0, 1.0 );
uniform vec4 line_color : source_color = vec4( 0.0, 1.0, 1.0, 1.0 );

float get_ratio_scan_line( float p )
{
	return max(
		-sin( mod( p, line_size ) / line_size * PI ) + line_width
	,	0.0
	) / line_width;
}

float hatch( vec2 src_uv, float time, float dir )
{
	float r = time * rotate_speed * dir;
	float c = cos( r );
	float s = sin( r );
	mat2 matr = mat2(
		vec2( c, s )
	,	vec2( -s, c )
	);
	vec2 uv = ( ( src_uv - vec2( 0.5, 0.5 ) ) * scale * time * 0.5 ) * matr;
	return clamp( 
		(
			get_ratio_scan_line( uv.x )
		+	get_ratio_scan_line( uv.y )
		) * sin( time * PI ) * 4.0
	,	0.0
	,	1.0
	);
}

void fragment( )
{
	float time = mod( TIME * speed, 0.3 ) / 0.3;
	float f = clamp(
		hatch( SCREEN_UV, time, 1.0 )
	+	hatch( SCREEN_UV, mod( time + 0.5, 1.0 ), -1.0 )
	,	0.0
	,	1.0
	);

	COLOR = mix( back_color, line_color, f );
}
```

### Steps to reproduce

Create a Minimal Project, attach the shader above to a ColorRect, with default presets in shader, switch project from Vulkan to GLES3, and attempt to run the project.

### Minimal reproduction project

_No response_

--- a/drivers/gles3/storage/material_storage.cpp
+++ b/drivers/gles3/storage/material_storage.cpp
@@ -1379,9 +1379,9 @@ MaterialStorage::MaterialStorage() {
 		actions.renames["POINT_SIZE"] = "gl_PointSize";
 
 		actions.renames["MODEL_MATRIX"] = "model_matrix";
-		actions.renames["CANVAS_MATRIX"] = "canvas_data.canvas_transform";
-		actions.renames["SCREEN_MATRIX"] = "canvas_data.screen_transform";
-		actions.renames["TIME"] = "canvas_data.time";
+		actions.renames["CANVAS_MATRIX"] = "canvas_transform";
+		actions.renames["SCREEN_MATRIX"] = "screen_transform";
+		actions.renames["TIME"] = "time";
 		actions.renames["PI"] = _MKSTR(Math_PI);
 		actions.renames["TAU"] = _MKSTR(Math_TAU);
 		actions.renames["E"] = _MKSTR(Math_E);
@@ -1399,7 +1399,7 @@ MaterialStorage::MaterialStorage() {
 		actions.renames["SPECULAR_SHININESS"] = "specular_shininess";
 		actions.renames["SCREEN_UV"] = "screen_uv";
 		actions.renames["SCREEN_TEXTURE"] = "screen_texture";
-		actions.renames["SCREEN_PIXEL_SIZE"] = "canvas_data.screen_pixel_size";
+		actions.renames["SCREEN_PIXEL_SIZE"] = "screen_pixel_size";
 		actions.renames["FRAGCOORD"] = "gl_FragCoord";
 		actions.renames["POINT_COORD"] = "gl_PointCoord";
 		actions.renames["INSTANCE_ID"] = "gl_InstanceIndex";

